---
title: "Cell_Chat_Mets"
author: "Alos Diallo"
date: "2024-08-07"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


```{r message=FALSE, warning=FALSE, include=FALSE}
library(Seurat)
library(CellChat)
library(futile.logger)
library(ggplot2)
library(stringr)
library(patchwork)
library(harmony)
library(readr)
library(viridis)
library(RCurl)
library(cowplot)
library(phateR)
library(compositions)
library(tidyverse)
library(dplyr)
library(hdf5r)
library(Matrix)
library(data.table)
library(ComplexHeatmap)
library(SingleR)
library(celldex)
library(SingleCellExperiment)
library(reticulate)
#library(loomR)
library(anndata)
library(DESeq2)
library(viridis)
library(textshaping)
library(EnhancedVolcano)
library(DataExplorer)
```


```{r}
options(future.globals.maxSize = 600 * 1024^2)
```


```{r}
combinded_data <- readRDS(file = "/Users/adiallo/Desktop/Thesis/Data_Documents/All_Data/Single_cell/Single_cell_samples/Single Cell/combined_data.rds")
```


```{r}

# Load the Seurat object from an RDS file
atlas <- readRDS("/Users/adiallo/Desktop/Thesis/Data_Documents/All_Data/Visium_samples/Visium_Data/atlas_data.rds")

# Normalize the atlas data using LogNormalize
atlas <- NormalizeData(atlas, normalization.method = "LogNormalize", scale.factor = 10000)
atlas <- FindVariableFeatures(atlas, selection.method = "vst", nfeatures = 2000)
atlas <- ScaleData(atlas)

atlas_data = atlas
```


```{r}
# Subset the combined data for metastatic cells
metastatic_data <- subset(combinded_data, subset = dataset_origin == "metastasis")

# Subset the combined data for non-metastatic cells
non_metastatic_data <- subset(combinded_data, subset = dataset_origin == "non_metastasis")
```

```{r}
colnames(metastatic_data@meta.data)
```
```{r}
# Define a mapping of original labels to consolidated labels
consolidated_labels <- c(
  "cB2 (B GC-like)" = "B cells",
  "cE02 (Stem/TA-like/Immature Goblet)" = "Tumor cells",
  "cE04 (Enterocyte 1)" = "Tumor cells",
  "cE03 (Stem/TA-like prolif)" = "Tumor cells",
  "cM01 (Monocyte)" = "Monocytes",
  "cS28 (GREM1+ CAF)" = "Fibroblasts",
  "cM02 (Macrophage-like)" = "Macrophages",
  "cS06 (Endo lymphatic)" = "Endothelial cells",
  "cP2 (Plasma IgG)" = "Plasma cells",
  "cS29 (MMP3+ CAF)" = "Fibroblasts",
  "cMA01 (Mast)" = "Mast cells",
  "cP1 (Plasma IgA)" = "Plasma cells",
  "cS33 (Schwann)" = "Schwann cells",
  "cTNI08 (CD4+ Treg)" = "Treg",
  "cE06 (Immature Goblet)" = "Tumor cells",
  "cS04 (Endo)" = "Endothelial cells",
  "cS17 (Pericyte)" = "Pericytes",
  "cS30 (CAF CCL8 Fibro-like)" = "Fibroblasts",
  "cS31 (CAF stem niche Fibro-like)" = "Fibroblasts",
  "cTNI13 (CD8+ T IL17+)" = "CD8+ T-cells",
  "cTNI17 (gd-like T)" = "CD8+ T-cells",
  "cE01 (Stem/TA-like)" = "Tumor cells",
  "cS27 (CXCL14+ CAF)" = "Fibroblasts",
  "cTNI01 (CD4+ IL7R+)" = "CD4+ T-cells",
  "cE05 (Enterocyte 2)" = "Tumor cells",
  "cB3 (B CD40+ GC-like)" = "B cells",
  "cM07 (pDC)" = "Dendritic cells",
  "cB1 (B IGD+IgM+)" = "B cells",
  "cTNI04 (CD4+ IL7R+CCL5+)" = "CD4+ T-cells",
  "cS26 (Myofibro)" = "Fibroblasts",
  "cS13 (Endo venous-like)" = "Endothelial cells",
  "cTNI18 (gd-like T PDCD1+)" = "T cells",
  "cTNI21 (PLZF+ T prolif)" = "T cells",
  "cS08 (Endo arterial-like)" = "Endothelial cells",
  "cP3 (Plasma IgG prolif)" = "Plasma cells",
  "cM09 (mregDC)" = "Dendritic cells",
  "cTNI14 (CD8+ CXCL13+)" = "CD8+ T-cells",
  "cTNI26 (ILC3)" = "Innate lymphoid cells",
  "cM10 (Granulocyte)" = "Granulocytes",
  "cM04 (DC2)" = "Dendritic cells",
  "cTNI05 (CD4+ IL17+)" = "CD4+ T-cells",
  "cTNI03 (CD4+ IL7R+HSP+)" = "CD4+ T-cells",
  "cS05 (Endo venous)" = "Endothelial cells",
  "cS19 (Pericyte)" = "Pericytes",
  "cTNI02 (CD4+ IL7R+SELL+)" = "CD4+ T-cells",
  "cTNI20 (PLZF+ T)" = "T cells",
  "cTNI10 (CD8+ IL7R+)" = "CD8+ T-cells",
  "cS09 (Endo)" = "Endothelial cells",
  "cTNI09 (CD4+ Treg prolif)" = "Treg",
  "cS16 (Pericyte)" = "Pericytes",
  "cTNI11 (CD8+GZMK+)" = "CD8+ T-cells",
  "cS12 (Endo)" = "Endothelial cells",
  "cE08 (Goblet)" = "Goblet cells",
  "cTNI06 (CD4+ TFH)" = "TFH cells",
  "cS32 (Smooth Muscle)" = "Smooth muscle cells",
  "cS14 (Endo)" = "Endothelial cells",
  "cS25 (Fibro CCL8+)" = "Fibroblasts",
  "cS22 (Fibro stem cell niche)" = "Fibroblasts",
  "cS10 (Endo tip cells)" = "Endothelial cells",
  "cE11 (Enteroendocrine)" = "Enteroendocrine cells",
  "cM05 (DC2 C1Q+)" = "Dendritic cells",
  "cTNI15 (CD8+ CXCL13+ HSP+)" = "CD8+ T-cells",
  "cTNI07 (CD4+ CXCL13+)" = "CD4+ T-cells",
  "cTNI16 (CD8+ CXCL13+ prolif)" = "CD8+ T-cells",
  "cTNI22 (cTNI22)" = "cTNI22",
  "cE10 (Tuft)" = "Tuft cells",
  "cS15 (Pericyte)" = "Pericytes",
  "cTNI12 (CD8+ IL7R+)" = "CD8+ T-cells",
  "cS20 (Pericyte prolif)" = "Pericytes",
  "cS18 (Pericyte)" = "Pericytes",
  # Adding the missing labels
  "cS11 (Endo proif)" = "Endothelial cells",
  "cM06 (DC IL22RA2)" = "Dendritic cells",
  "cS02 (Endo capillary)" = "Endothelial cells",
  "cS01 (Endo arterial)" = "Endothelial cells",
  "cE09 (Best4)" = "Goblet cells",
  "cS03 (Endo capillary)" = "Endothelial cells",
  "cS07 (Endo capillary-like)" = "Endothelial cells",
  "cS24 (Fibro BMP-producing)" = "Fibroblasts",
  "cM08 (AS-DC)" = "Dendritic cells",
  "cTNI23 (NK CD16A+)" = "NK cells",
  "cS21 (Fibro stem cell niche)" = "Fibroblasts",
  "cTNI24 (NK GZMK+)" = "NK cells",
  "cM03 (DC1)" = "Dendritic cells",
  "cE07 (Goblet/Enterocyte)" = "Goblet cells"
)

# Apply the consolidated labels to the metadata
metastatic_data@meta.data$predicted.id <- consolidated_labels[metastatic_data@meta.data$predicted.id]
non_metastatic_data@meta.data$predicted.id <- consolidated_labels[non_metastatic_data@meta.data$predicted.id]

# Extract the updated predicted.id
mets_labels <- metastatic_data@meta.data$predicted.id
nonmets_labels <- non_metastatic_data@meta.data$predicted.id

# Ensure that the names of the annotations match the cell barcodes
names(mets_labels) <- rownames(metastatic_data@meta.data)
# Ensure that the names of the annotations match the cell barcodes
names(nonmets_labels) <- rownames(non_metastatic_data@meta.data)

# Create the updated meta data frame
mets_meta = NULL
mets_meta <- data.frame(labels = mets_labels, row.names = names(mets_labels))
nonmets_meta = NULL
nonmets_meta <- data.frame(labels = nonmets_labels, row.names = names(nonmets_labels))

# Extract the normalized data matrix using the correct method for Seurat v5.0.0
mets_data.input <- metastatic_data[["RNA"]]$data
non_mets_data.input <- non_metastatic_data[["RNA"]]$data

```

Non of this code block works and breaks the code
```{r}
# Convert the labels to a factor (if not already)
mets_meta$labels <- as.factor(mets_meta$labels)

# Drop unused levels in the factor
mets_meta$labels <- droplevels(mets_meta$labels)
# If you want to assign NA values to a specific label instead, you can do so:
mets_meta$labels[is.na(mets_meta$labels)] <- "Unknown"

# Example: Merge `cTNI22` with `T cells`
mets_meta$labels <- recode(mets_meta$labels, `cTNI22` = "T cells")

# Drop unused levels
mets_meta$labels <- droplevels(mets_meta$labels)

# Exclude groups with fewer than 10 cells
valid_labels <- names(which(table(mets_meta$labels) >= 10))
mets_meta <- mets_meta[mets_meta$labels %in% valid_labels, , drop = FALSE]  # Keep as a data frame

# Drop unused levels
mets_meta$labels <- droplevels(mets_meta$labels)


```


```{r}
cellchat_mets = NULL
cellchat_non_Mets = NULL
# Create CellChat object
cellchat_mets <- createCellChat(object = mets_data.input, meta = mets_meta, group.by = "labels")
# Create CellChat object
cellchat_non_Mets <- createCellChat(object = non_mets_data.input, meta = nonmets_meta, group.by = "labels")
```
```{r}
CellChatDB <- CellChatDB.human 
```

```{r}
# use a subset of CellChatDB for cell-cell communication analysis
CellChatDB.use <- subsetDB(CellChatDB, search = "Cell-Cell Contact", key = "annotation") # use Secreted Signaling

# Only uses the Secreted Signaling from CellChatDB v1
#  CellChatDB.use <- subsetDB(CellChatDB, search = list(c("Secreted Signaling"), c("CellChatDB v1")), key = c("annotation", "version"))

# use all CellChatDB except for "Non-protein Signaling" for cell-cell communication analysis
# CellChatDB.use <- subsetDB(CellChatDB)


# use all CellChatDB for cell-cell communication analysis
# CellChatDB.use <- CellChatDB # simply use the default CellChatDB. We do not suggest to use it in this way because CellChatDB v2 includes "Non-protein Signaling" (i.e., metabolic and synaptic signaling). 

# set the used database in the object
cellchat_mets@DB <- CellChatDB.use
cellchat_non_Mets@DB <- CellChatDB.use
```



```{r}
options(future.globals.maxSize = 4000 * 1024^2)
#devtools::install_github("immunogenomics/presto")
#Needs presto package
# subset the expression data of signaling genes for saving computation cost
cellchat_mets <- subsetData(cellchat_mets) # This step is necessary even if using the whole database
future::plan("multisession", workers = 4) # do parallel
cellchat_mets <- identifyOverExpressedGenes(cellchat_mets)
cellchat_mets <- identifyOverExpressedInteractions(cellchat_mets)
#> The number of highly variable ligand-receptor pairs used for signaling inference is 692

cellchat_non_Mets <- subsetData(cellchat_non_Mets) # This step is necessary even if using the whole database
future::plan("multisession", workers = 4) # do parallel
cellchat_non_Mets <- identifyOverExpressedGenes(cellchat_non_Mets)
cellchat_non_Mets <- identifyOverExpressedInteractions(cellchat_non_Mets)

#> [1] 13.20763
# project gene expression data onto PPI (Optional: when running it, USER should set `raw.use = FALSE` in the function `computeCommunProb()` in order to use the projected data)
#cellchat <- CellChat::projectData(cellchat, PPI.human)
```


```{r}
cellchat_mets <- computeCommunProb(cellchat_mets, type = "triMean")
cellchat_non_Mets <- computeCommunProb(cellchat_non_Mets, type = "triMean")
```
Filter out cells
```{r}
cellchat_mets <- filterCommunication(cellchat_mets, min.cells = 3)
cellchat_non_Mets <- filterCommunication(cellchat_non_Mets, min.cells = 10)
```


```{r}
cellchat_mets <- computeCommunProbPathway(cellchat_mets)
cellchat_non_Mets <- computeCommunProbPathway(cellchat_non_Mets)
```


```{r}
cellchat_mets <- aggregateNet(cellchat_mets)
cellchat_non_Mets <- aggregateNet(cellchat_non_Mets)
```

```{r}
groupSize <- as.numeric(table(cellchat_mets@idents))
par(mfrow = c(1,2), xpd=TRUE)
netVisual_circle(cellchat_mets@net$count, vertex.weight = groupSize, weight.scale = T, label.edge= F, title.name = "Number of interactions")
netVisual_circle(cellchat_mets@net$weight, vertex.weight = groupSize, weight.scale = T, label.edge= F, title.name = "Interaction weights/strength")
```
```{r}
mat <- cellchat_mets@net$weight
par(mfrow = c(3,4), xpd=TRUE)
for (i in 1:nrow(mat)) {
  mat2 <- matrix(0, nrow = nrow(mat), ncol = ncol(mat), dimnames = dimnames(mat))
  mat2[i, ] <- mat[i, ]
  netVisual_circle(mat2, vertex.weight = groupSize, weight.scale = T, edge.weight.max = max(mat), title.name = rownames(mat)[i])
}
```
```{r}
pathways.show <- c("DESMOSOME") 
# Hierarchy plot
# Here we define `vertex.receive` so that the left portion of the hierarchy plot shows signaling to fibroblast and the right portion shows signaling to immune cells 
vertex.receiver = seq(1,4) # a numeric vector. 
netVisual_aggregate(cellchat_non_Mets, signaling = pathways.show,  vertex.receiver = vertex.receiver)
# Circle plot
par(mfrow=c(1,2))
netVisual_aggregate(cellchat_non_Mets, signaling = pathways.show, layout = "circle")

netVisual_aggregate(cellchat_mets, signaling = pathways.show, layout = "circle")


```
```{r}
cellchat_mets@dr$umap <- Embeddings(metastatic_data, "umap")
cellchat_non_Mets@dr$umap <- Embeddings(non_metastatic_data, "umap")
```

