---
title: "Sabin_DH_EWAS"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# visualize PCA for cell mixtures
```{r}
Z <- prcomp(as.matrix(t(DHColon_HTD)))
Z <- as.data.frame(Z$rotation[,c("PC1","PC2")])
colnames(Z) <- c("x","y")
Z$patient <- rownames(Z)
Z <- merge(Z,covar.data,on="patient")

ggplot(Z,aes(x,y,color=factor(MLH1))) + geom_point() + theme_classic() 

ggplot(Z,aes(x,y,color=factor(any_mets))) + geom_point() + theme_classic()
```

# visualize PCA for beta values
```{r}
Z <- prcomp(as.matrix(beta.val))
Z <- as.data.frame(Z$rotation[,c("PC1","PC2")])
colnames(Z) <- c("x","y")
Z$patient <- rownames(Z)
Z <- merge(Z,covar.data,on="patient")

ggplot(Z,aes(x,y,color=factor(MLH1))) + geom_point()  + theme_classic()

ggplot(Z,aes(x,y,color=factor(ln_only))) + geom_point() + facet_grid(~MLH1) + theme_classic()
```

# Run EWAS, with example cell type proportion adjustment
```{r}
library(limma)
M.val <- ifelse(beta.val.expanded>=1,1-1E-6,ifelse(beta.val.expanded<=0,1E-6,beta.val.expanded))
M.val <- log(M.val)-log(1-M.val)

# cell type adjusted
X.cell <- model.matrix(~any_mets+age+sex+MLH1+Tumor+Epithelial+Endothelial+Stromal+Treg,data=df.cell) # note: add more cell types!
# %>% filter(!Distant_Mets)

# no cell type adjustment
X.nocell <- model.matrix(~any_mets+age+sex+MLH1,data=df.cell)

# hitimed cell adjustments
X.hitimed <- model.matrix(~any_mets+age+sex+MLH1+Tumor+Epithelial+Endothelial+Stromal+Bmem+CD8mem+Treg+DC,data=df.cell)


block <- NULL

# !!
# corfit <- duplicateCorrelation(M.val[1:350000,], X.nocell, block=block, ndups=2)
corfit <- readRDS("corfit.rds")
fit <- lmFit(M.val,X.cell,block=block,correlation=corfit$consensus)
eb <- eBayes(fit)
res <- topTable(eb,coef=2,adjust="BH",n=Inf)
res[1:10,]

# hitimed EWAS
#corfit2 <- duplicateCorrelation(M.val[1:350000,], X.hitimed, block=block, ndups=2)
corfit2 <- readRDS("corfit_adjusted.rds")
fit2 <- lmFit(M.val,X.hitimed,block=block,correlation=corfit2$consensus)
eb2 <- eBayes(fit2)
res2 <- topTable(eb2,coef=2,adjust="BH",n=Inf)
res2[1:10,]



```

# Visualize Volcano
```{r}
library(IlluminaHumanMethylationEPICanno.ilm10b4.hg19)
library(EnhancedVolcano)
strip.names<-function(x){
  return(sapply(strsplit(x,";"), function(x) ifelse(length(x)>0,getElement(x,1),"")))
}

cpg2gene<-getAnnotation(IlluminaHumanMethylationEPICanno.ilm10b4.hg19)[,c("Name","UCSC_RefGene_Name")]
cpg2gene<-setNames(strip.names(cpg2gene$UCSC_RefGene_Name),cpg2gene$Name)
res$gene<-gene.names<-cpg2gene[rownames(res)]
res2$gene<-gene.names2 <- cpg2gene[rownames(res2)]



EnhancedVolcano(res,
    title = "DHMC EWAS No Adjustment",
    lab = gene.names,
    x = 'logFC',
    y = 'P.Value',
    pCutoff = 0.05/100,
    FCcutoff = 1.2,
    drawConnectors = T,
    boxedLabels = T,
    widthConnectors = 0.25,
    max.overlaps=30)
#ggsave("DH_EWAS_noadjust.png", device = "png", path = "./Sabin/")


EnhancedVolcano(res2,
    title = "DHMC EWAS HiTIMED Adjustments",
    lab = gene.names2,
    x = 'logFC',
    y = 'P.Value',
    pCutoff = 0.05/100,
    FCcutoff = 1.2,
    drawConnectors = T,
    boxedLabels = T,
    widthConnectors = 0.25,
    max.overlaps=30)
#ggsave("DH_EWAS_adjust.png", device = "png", path = "./Sabin/")

```

# Pathway Analysis, WIP
```{r}
library(missMethyl)
length(rownames(res %>% filter(P.Value<=0.0005)))

gst <- gometh(sig.cpg = rownames(res %>% filter(P.Value<=0.005)), all.cpg = rownames(res), collection = "KEGG", plot.bias = TRUE, prior.prob = TRUE)
gst %>% filter(P.DE<=0.05)
```

# Run CellDMC, example with 10000 CpGs
```{r}
library(EpiDISH)
ma.B.new<-beta.val.filtered[1:10000,]
cell.dmc.res<-CellDMC(
  ma.B.new,
  pheno.v=covar.data$ln_only,
  frac.m=as.matrix(DHColon_HTD)[,colMeans(DHColon_HTD>5)>=0.05]/100,
  adjPMethod = "fdr",
  adjPThresh = 0.1,
  cov.mod = model.matrix(~.,df.cell %>% select(age,sex,MLH1))[,-1],
  sort = FALSE,
  mc.cores = 1
)
dmc.interaction.results<-cell.dmc.res$dmct[apply(cell.dmc.res$dmct,1,function(x) any(x!=0)),]
dmc.interaction.results
```

# Cell DMC Results
```{r}
col_colors<-covar.data %>% select(MLH1,age,sex,any_mets,Distant_Mets,ln_only)
row_colors<-dmc.interaction.results
rownames(row_colors)<-rownames(dmc.interaction.results)
pheatmap(ma.B.new[rownames(dmc.interaction.results),],
         annotation_row=as.data.frame(row_colors),
         annotation_col=as.data.frame(col_colors),
         show_rownames = F, show_colnames = F, cellwidth=5, cellheight=15, fontsize=6,treeheight_col=0,treeheight_row=0,legend=F,annotation_legend=F)
```


# load TCGA data
```{r}
# library(ChAMP)
median.impute<-T
cell_type_names<-c("Tumor","Endothelial","Epithelial","Stromal","Bnv","Bmem","CD4nv","CD4mem","Treg","CD8nv","CD8mem","Mono","DC","NK","Bas","Eos","Neu")
tcga.data<-readRDS("TCGA_data.rds")
beta.val<-t(as.matrix(tcga.data$dnam))
if (median.impute){
  beta.val<-apply(beta.val,1,function(x){
    x[is.na(x)]<-median(x,na.rm=T)
    x
  })
  beta.val<-t(beta.val)
} else {
  beta.val<-champ.impute(beta.val)
}
tcga.data$pheno$ln_mets<-tcga.data$pheno$ajcc_pathologic_n!="N0"
# seems like all patients have LN METS, hmm... will investigate
cell.type.limit.df<-tcga.data$cell_types$COAD_Deconv6 %>% select(sex,age,cell_type_names)#select(gender, ajcc_pathologic_n, ajcc_pathologic_m, ln_mets)
pheno<-cbind(tcga.data$pheno,cell.type.limit.df)
pheno$patient<-rownames(pheno)
msi_results<-readRDS("msi_tcga.rds")
msi_results$patient<-msi_results$bcr_patient_barcode
msi_results$msi<-msi_results$mononucleotide_and_dinucleotide_marker_panel_analysis_status
pheno<-pheno %>% merge(msi_results,by="patient",all.x=T)
rownames(pheno)<-pheno$patient
sample.filter<-(rowSums(is.na(pheno %>% select(cell_type_names)))==0)
beta.val<-beta.val[,sample.filter]
pheno<-pheno[sample.filter,]
pheno<-pheno[colnames(beta.val),]

mad.vals<-apply(beta.val,1,mad)
beta.val.filtered<-beta.val[names(sort(rank(-mad.vals))[1:10000]),]
```

# plot cell types
```{r}
pheno %>% pivot_longer(cols=cell_type_names) %>% ggplot(aes(x=factor(name),y=value,color=factor(ln_mets))) + geom_boxplot()  + theme_classic()+ theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=0))

pheatmap(pheno %>% select(cell_type_names),show_colnames=T,show_rownames = F,scale = 'column', annotation_row=pheno %>% select(ln_mets,msi) %>% mutate(ln_mets=as.numeric(ln_mets)))
```

# plot beta values
```{r}
pheatmap(beta.val.filtered[1:10000,],show_colnames=F,show_rownames = F, annotation_col=pheno %>% select(ln_mets,msi) %>% mutate(ln_mets=as.numeric(ln_mets)),treeheight_row=0)
```

# visualize PCA for cell mixtures
```{r}
Z <- prcomp(as.matrix(t(pheno %>% select(cell_type_names))))
Z<-as.data.frame(Z$rotation[,c("PC1","PC2")])
colnames(Z)<-c("x","y")
Z$patient<-rownames(Z)
Z<-merge(Z,pheno,on="patient")

ggplot(Z,aes(x,y,color=factor(ln_mets))) + geom_point() + facet_grid(~as.factor(msi)) + theme_classic() 

```

# visualize PCA for beta values
```{r}
Z <- prcomp(as.matrix(beta.val.filtered))
Z<-as.data.frame(Z$rotation[,c("PC1","PC2")])
colnames(Z)<-c("x","y")
Z$patient<-rownames(Z)
Z<-merge(Z,pheno,on="patient")

ggplot(Z,aes(x,y,color=factor(ln_mets))) + geom_point()  + facet_grid(~as.factor(msi)) +  theme_classic()

```

# Run EWAS, with example cell type proportion adjustment
```{r}
library(limma)
M.val.filtered <- ifelse(beta.val.filtered>=1,1-1E-6,ifelse(beta.val.filtered<=0,1E-6,beta.val.filtered))
M.val.filtered <- log(M.val.filtered)-log(1-M.val.filtered)
X.cell<-model.matrix(~ln_mets+age+gender+msi+Tumor+Epithelial+Endothelial+Stromal+Treg,data=pheno) # note: add more cell types!
X.nocell<-model.matrix(~ln_mets+age+gender+msi,data=pheno)
block<-NULL
corfit <- duplicateCorrelation(M.val.filtered[1:10000,], X.cell, block=block, ndups=2)
fit<-lmFit(M.val.filtered,X.cell,block=block,correlation=corfit$consensus)
eb <- eBayes(fit)
res <- topTable(eb,coef=2,adjust="BH",n=Inf)
res[1:10,]
```

# Visualize Volcano
```{r}
library(IlluminaHumanMethylationEPICanno.ilm10b4.hg19)
library(EnhancedVolcano)
strip.names<-function(x){
  return(sapply(strsplit(x,";"), function(x) ifelse(length(x)>0,getElement(x,1),"")))
}

cpg2gene<-getAnnotation(IlluminaHumanMethylationEPICanno.ilm10b4.hg19)[,c("Name","UCSC_RefGene_Name")]
cpg2gene<-setNames(strip.names(cpg2gene$UCSC_RefGene_Name),cpg2gene$Name)
gene.names<-cpg2gene[rownames(res)]

EnhancedVolcano(res,
    lab = gene.names,
    x = 'logFC',
    y = 'P.Value',
    pCutoff = 0.05/10000,
    FCcutoff = 0.5,
    drawConnectors = T,
    widthConnectors = 0.25)
```

# Pathway Analysis, WIP
```{r}
library(missMethyl)
gst <- gometh(sig.cpg = rownames(res %>% filter(P.Value<=0.001)), all.cpg = rownames(res), collection = "KEGG", plot.bias = TRUE, prior.prob = TRUE)
gst %>% filter(P.DE<=0.05)
```

# Run CellDMC, example with 10000 CpGs
```{r}
library(EpiDISH)
select<-dplyr::select
ma.B.new<-beta.val.filtered[1:10000,]
cell.dmc.res<-CellDMC(
  ma.B.new,
  pheno.v=pheno$ln_mets,
  frac.m=as.matrix(pheno %>% select(cell_type_names))[,colMeans((pheno %>% select(cell_type_names))>5)>=0.05]/100,
  adjPMethod = "fdr",
  adjPThresh = 5e-2,
  cov.mod = model.matrix(~.,pheno %>% select(age,gender))[,-1],
  sort = FALSE,
  mc.cores = 1
)
dmc.interaction.results<-cell.dmc.res$dmct[apply(cell.dmc.res$dmct,1,function(x) any(x!=0)),]
dmc.interaction.results
```

# Cell DMC Results
```{r}
col_colors<-pheno %>% select(age,gender,ln_mets) %>% mutate(ln_mets=as.numeric(ln_mets))
row_colors<-dmc.interaction.results
rownames(row_colors)<-rownames(dmc.interaction.results)
pheatmap(ma.B.new[rownames(dmc.interaction.results),],
         annotation_row=as.data.frame(row_colors),
         annotation_col=as.data.frame(col_colors),
         show_rownames = F, show_colnames = F, cellwidth=2, cellheight=15, fontsize=6,treeheight_col=0,treeheight_row=0,legend=F,annotation_legend=F)
```

# OLD CODE
```{r}
# TCGA data here
# cell types, similar covariates, matching beta, etc. metaanalysis
# derive interactions to be tested in DH data
# if (F){
  # library(reticulate)
  # options(reticulate.conda_binary = "/dartfs-hpc/rc/lab/V/VaickusL/anaconda_installation/bin/conda")
  # use_condaenv("detectron")
  # pd<-import("pandas")
  # tcga.data<-pd$read_pickle("../tcga_colon_dnam/dnam_data_TCGA_HiTIMED.pkl")
  # for (nm in names(tcga.data$cell_types)){
  #   tcga.data$cell_types[[nm]]<-py_to_r(tcga.data$cell_types[[nm]])
  # }
  # saveRDS(tcga.data,"TCGA_data.rds")
# }
```

# get MSI data
```{r}
library(TCGAbiolinks)
query <- GDCquery(project = "TCGA-COAD", 
                  data.category = "Other",
                  legacy = TRUE,
                  access = "open",
                  data.type = "Auxiliary test",
                  )  
GDCdownload(query)
msi_results <- GDCprepare_clinic(query, "msi")
rownames(msi_results)<-msi_results$bcr_patient_barcode
saveRDS(msi_results,"msi_tcga.rds")
# TODO: save and add to pheno after matching columns
```


